src/future/value_initializer.rs

```diff
@@ -35,70 +36,99 @@ where
+        use std::panic::{resume_unwind, AssertUnwindSafe};
         use InitResult::*;
 
-        let type_id = TypeId::of::<E>();
-        let waiter = Arc::new(RwLock::new(None));
-        let mut lock = waiter.write().await;
+        loop {
+            let waiter = Arc::new(RwLock::new(None));
+            let mut lock = waiter.write().await;
 
-        match self.try_insert_waiter(&key, type_id, &waiter) {
-            None => {
-                // Our waiter was inserted. Let's resolve the init future.
-                match init.await {
-                    Ok(value) => {
-                        *lock = Some(Ok(value.clone()));
-                        Initialized(value)
-                    }
-                    Err(e) => {
-                        let err: ErrorObject = Arc::new(e);
-                        *lock = Some(Err(Arc::clone(&err)));
-                        self.remove_waiter(&key, type_id);
-                        InitErr(err.downcast().unwrap())
+            match self.try_insert_waiter(key, type_id, &waiter) {
+                None => {
+                    // Our waiter was inserted. Let's resolve the init future.
+                    // Catching panic is safe here as we do not try to resolve the future again.
+                    match AssertUnwindSafe(init).catch_unwind().await {
+                        // Resolved.
+                        Ok(value) => return post_init(key, value, &mut lock),
+                        // Panicked.
+                        Err(payload) => {
+                            *lock = None;
+                            // Remove the waiter so that others can retry.
+                            self.remove_waiter(key, type_id);
+                            resume_unwind(payload);
+                        } // The lock will be unlock here.
                     }
                 }
-            }
-            Some(res) => {
-                // Somebody else's waiter already exists. Drop our write lock and wait
-                // for a read lock to become available.
-                std::mem::drop(lock);
-                match &*res.read().await {
-                    Some(Ok(value)) => ReadExisting(value.clone()),
-                    Some(Err(e)) => InitErr(Arc::clone(e).downcast().unwrap()),
-                    None => unreachable!(),
+                Some(res) => {
+                    // Somebody else's waiter already exists. Drop our write lock and wait
+                    // for a read lock to become available.
+                    std::mem::drop(lock);
+                    match &*res.read().await {
+                        Some(Ok(value)) => return ReadExisting(value.clone()),
+                        Some(Err(e)) => return InitErr(Arc::clone(e).downcast().unwrap()),
+                        // None means somebody else's init future has been panicked.
+                        // Retry from the beginning.
+                        None => continue,
```

Comment:
Just to be safer, we should avoid to create a busy, infinite loop here. Maybe we can do some of the followings:

- Add a retry limit. If the number of retries has reached to the limit, do panic with some helpful message.
- If it is still getting `None` after some retries, it probably means other `init` futures are also panicking. But is there any other reason? e.g. removing the waiter is failing? If so, is there anything we can do here?
    - Adding a short sleep between retries to avoid busy loop?
    - Try to remove the waiter here?

I think we just want to do the first one for now, and see what happens.

Reply:
> Add a retry limit. If the number of retries has reached to the limit, do panic with some helpful message.

Added the retry limit: https://github.com/moka-rs/moka/pull/46/commits/af44363001962d92b425f6a139d2dc82f0d242ce